datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -153,103 +264,31 @@ where
-                        // and will therefore return false positives
-                        debug!(
-                            logger,
-                            "Starting request for logs in block range [{},{}]",
-                            chunk_offset,
-                            chunk_end
-                        );
-                        let log_filter = log_filter.clone();
-                        let chunk_future = eth_adapter
-                            .logs_with_sigs(
-                                &logger,
-                                chunk_offset,
-                                chunk_end,
-                                addresses.clone(),
-                                event_sigs.clone(),
-                            )
-                            .map(move |logs| {
-                                logs.into_iter()
-                                    // Filter out false positives
-                                    .filter(move |log| log_filter.matches(log))
-                                    .collect()
-                            });
-
-                        // Save future for later
-                        chunk_futures.push(Box::new(chunk_future)
-                            as Box<Future<Item = Vec<Log>, Error = _> + Send>);
-
-                        // If last chunk, will push offset past `to`. That's fine.
-                        chunk_offset += *LOG_STREAM_CHUNK_SIZE_IN_BLOCKS;
-
-                        if is_last_chunk {
-                            break;
-                        }
-                    }
-                }
-
-                // Combine chunk futures into one future (Vec<Log>, u64)
-                Some(stream::futures_ordered(chunk_futures).collect().map(
-                    move |chunks: Vec<Vec<Log>>| {
-                        let flattened = chunks.into_iter().flat_map(|v| v).collect::<Vec<Log>>();
-                        (flattened, chunk_offset)
-                    },
-                ))
-            } else {
-                None
+        stream::unfold(from, move |start| {
+            if start > to {
+                return None;
             }
+            let end = if start < *LOG_STREAM_FAST_SCAN_END {
+                (start + 100_000 - 1).min(to).min(*LOG_STREAM_FAST_SCAN_END)
+            } else {
+                (start + 1_000 - 1).min(to)
+            };
+            let new_start = end + 1;
+
+            debug!(logger, "Requesting logs for blocks [{}, {}]", start, end);
+
+            let log_filter = log_filter.clone();
+            Some(
+                eth.logs_with_sigs(&logger, start, end, addresses.clone(), event_sigs.clone())
+                    .map(move |logs| {
+                        logs.into_iter()
+                            .filter(move |log| log_filter.matches(log))
+                            .collect()
+                    })
+                    .map(move |logs: Vec<Log>| (logs, new_start)),
```

Comment:
I'd move this `Vec<Log>` up and make it `collect::<Vec<Log>>`. It feels that's more obvious.